From 6506f28f5b491dc4c3ff2a8087930ae93c82d676 Mon Sep 17 00:00:00 2001
From: Asmitha Karunanithi <asmitk01@in.ibm.com>
Date: Fri, 7 Feb 2025 11:31:13 -0600
Subject: [PATCH 1/2] Add PCIeSlots Redfish Endpoint for Chassis

This commit introduces support for retrieving PCIe slot information
from the inventory and exposing it via the /Chassis/<id>/PCIeSlots
Redfish endpoint.

Upstream-Status: Pending

Tested By:
'''
GET https://${bmc}/redfish/v1/Chassis/<ChassisId>
    "PCIeSlots": {
        "@odata.id": "/redfish/v1/Chassis/IBM_System1_Baseboard/PCIeSlots"
    }

GET https://${bmc}/redfish/v1/Chassis/<ChassisId>/PCIeSlots
    "Slots": [
        {
          "Lanes": 4,
          "Location": {
            "PartLocation": {
              "ServiceLabel": "..."
            }
          },
          "PCIeType": "...",
          "SlotType": "...",
          "Status": {
            "State": "<Enabled/Absent>"
          }
        },
        {...}
    ]
'''

Redfish Validator Passed.

Change-Id: If9200559359c721aa67775d65f2dc9d3bf51a079
Signed-off-by: Asmitha Karunanithi <asmitk01@in.ibm.com>
---
 redfish-core/lib/chassis.hpp    |   2 +
 redfish-core/lib/pcie_slots.hpp | 198 +++++++++++++++++++-------------
 redfish-core/src/redfish.cpp    |   2 +
 3 files changed, 121 insertions(+), 81 deletions(-)

diff --git a/redfish-core/lib/chassis.hpp b/redfish-core/lib/chassis.hpp
index 90e3b60e83..a0fa87cb45 100644
--- a/redfish-core/lib/chassis.hpp
+++ b/redfish-core/lib/chassis.hpp
@@ -567,6 +567,8 @@ inline void handleChassisGetSubTree(
             .jsonValue["Actions"]["#Chassis.Reset"]["@Redfish.ActionInfo"] =
             boost::urls::format("/redfish/v1/Chassis/{}/ResetActionInfo",
                                 chassisId);
+        asyncResp->res.jsonValue["PCIeSlots"]["@odata.id"] =
+            boost::urls::format("/redfish/v1/Chassis/{}/PCIeSlots", chassisId);
         dbus::utility::getAssociationEndPoints(
             path + "/drive",
             [asyncResp, chassisId](const boost::system::error_code& ec3,
diff --git a/redfish-core/lib/pcie_slots.hpp b/redfish-core/lib/pcie_slots.hpp
index e658e3c2cf..9bd0a717d9 100644
--- a/redfish-core/lib/pcie_slots.hpp
+++ b/redfish-core/lib/pcie_slots.hpp
@@ -12,7 +12,6 @@
 #include "logging.hpp"
 #include "query.hpp"
 #include "registries/privilege_registry.hpp"
-#include "utils/dbus_utils.hpp"
 #include "utils/pcie_util.hpp"
 
 #include <asm-generic/errno.h>
@@ -22,8 +21,8 @@
 #include <boost/url/format.hpp>
 #include <nlohmann/json.hpp>
 #include <sdbusplus/message/native_types.hpp>
-#include <sdbusplus/unpack_properties.hpp>
 
+#include <algorithm>
 #include <array>
 #include <cstddef>
 #include <functional>
@@ -32,101 +31,116 @@
 #include <string>
 #include <string_view>
 #include <utility>
+#include <variant>
 
 namespace redfish
 {
 
-inline void onPcieSlotGetAllDone(
-    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
-    const boost::system::error_code& ec,
-    const dbus::utility::DBusPropertiesMap& propertiesList)
+inline void processPcieSlot(
+    const dbus::utility::DBusInterfacesMap& interfacesList,
+    nlohmann::json& slots)
 {
-    if (ec)
-    {
-        BMCWEB_LOG_ERROR("Can't get PCIeSlot properties!");
-        messages::internalError(asyncResp->res);
-        return;
-    }
-
-    nlohmann::json& slots = asyncResp->res.jsonValue["Slots"];
+    nlohmann::json slot = nlohmann::json::object();
+    bool slotHasData = false;
 
-    nlohmann::json::array_t* slotsPtr =
-        slots.get_ptr<nlohmann::json::array_t*>();
-    if (slotsPtr == nullptr)
+    for (const auto& [interface, properties] : interfacesList)
     {
-        BMCWEB_LOG_ERROR("Slots key isn't an array???");
-        messages::internalError(asyncResp->res);
-        return;
-    }
-
-    nlohmann::json::object_t slot;
-
-    const std::string* generation = nullptr;
-    const size_t* lanes = nullptr;
-    const std::string* slotType = nullptr;
-    const bool* hotPluggable = nullptr;
-
-    const bool success = sdbusplus::unpackPropertiesNoThrow(
-        dbus_utils::UnpackErrorPrinter(), propertiesList, "Generation",
-        generation, "Lanes", lanes, "SlotType", slotType, "HotPluggable",
-        hotPluggable);
-
-    if (!success)
-    {
-        messages::internalError(asyncResp->res);
-        return;
-    }
-
-    if (generation != nullptr)
-    {
-        std::optional<pcie_device::PCIeTypes> pcieType =
-            pcie_util::redfishPcieGenerationFromDbus(*generation);
-        if (!pcieType)
-        {
-            BMCWEB_LOG_WARNING("Unknown PCIe Slot Generation: {}", *generation);
-        }
-        else
+        if (interface == "xyz.openbmc_project.Inventory.Item.PCIeSlot")
         {
-            if (*pcieType == pcie_device::PCIeTypes::Invalid)
+            auto it = std::find_if(
+                properties.begin(), properties.end(),
+                [](const auto& pair) { return pair.first == "Generation"; });
+            if (it != properties.end())
             {
-                messages::internalError(asyncResp->res);
-                return;
+                const std::string* generation =
+                    std::get_if<std::string>(&it->second);
+                if (generation != nullptr)
+                {
+                    std::optional<pcie_device::PCIeTypes> pcieType =
+                        pcie_util::redfishPcieGenerationFromDbus(*generation);
+                    if (pcieType &&
+                        *pcieType != pcie_device::PCIeTypes::Invalid)
+                    {
+                        slot["PCIeType"] = *pcieType;
+                        slotHasData = true;
+                    }
+                }
             }
-            slot["PCIeType"] = *pcieType;
-        }
-    }
 
-    if (lanes != nullptr && *lanes != 0)
-    {
-        slot["Lanes"] = *lanes;
-    }
+            it = std::find_if(properties.begin(), properties.end(),
+                              [](const auto& pair) {
+                                  return pair.first == "Lanes";
+                              });
+            if (it != properties.end())
+            {
+                const size_t* lanes = std::get_if<size_t>(&it->second);
+                if (lanes != nullptr)
+                {
+                    slot["Lanes"] = *lanes;
+                    slotHasData = true;
+                }
+            }
 
-    if (slotType != nullptr)
-    {
-        std::optional<pcie_slots::SlotTypes> redfishSlotType =
-            pcie_util::dbusSlotTypeToRf(*slotType);
-        if (!redfishSlotType)
+            it = std::find_if(properties.begin(), properties.end(),
+                              [](const auto& pair) {
+                                  return pair.first == "SlotType";
+                              });
+            if (it != properties.end())
+            {
+                const std::string* slotType =
+                    std::get_if<std::string>(&it->second);
+                if (slotType != nullptr)
+                {
+                    std::optional<pcie_slots::SlotTypes> redfishSlotType =
+                        pcie_util::dbusSlotTypeToRf(*slotType);
+                    if (redfishSlotType &&
+                        *redfishSlotType != pcie_slots::SlotTypes::Invalid)
+                    {
+                        slot["SlotType"] = *redfishSlotType;
+                        slotHasData = true;
+                    }
+                }
+            }
+        }
+        else if (interface ==
+                 "xyz.openbmc_project.Inventory.Decorator.LocationCode")
         {
-            BMCWEB_LOG_WARNING("Unknown PCIe Slot Type: {}", *slotType);
+            auto it = std::find_if(
+                properties.begin(), properties.end(),
+                [](const auto& pair) { return pair.first == "LocationCode"; });
+            if (it != properties.end())
+            {
+                const std::string* locationCode =
+                    std::get_if<std::string>(&it->second);
+                if (locationCode != nullptr)
+                {
+                    slot["Location"]["PartLocation"]["ServiceLabel"] =
+                        *locationCode;
+                    slotHasData = true;
+                }
+            }
         }
-        else
+        else if (interface == "xyz.openbmc_project.Inventory.Item")
         {
-            if (*redfishSlotType == pcie_slots::SlotTypes::Invalid)
+            auto it = std::find_if(
+                properties.begin(), properties.end(),
+                [](const auto& pair) { return pair.first == "Present"; });
+            if (it != properties.end())
             {
-                BMCWEB_LOG_ERROR("Unknown PCIe Slot Type: {}", *slotType);
-                messages::internalError(asyncResp->res);
-                return;
+                const bool* present = std::get_if<bool>(&it->second);
+                if (present != nullptr)
+                {
+                    slot["Status"]["State"] = *present ? "Enabled" : "Absent";
+                    slotHasData = true;
+                }
             }
-            slot["SlotType"] = *redfishSlotType;
         }
     }
 
-    if (hotPluggable != nullptr)
+    if (slotHasData)
     {
-        slot["HotPluggable"] = *hotPluggable;
+        slots.push_back(slot);
     }
-
-    slots.emplace_back(std::move(slot));
 }
 
 inline void onMapperAssociationDone(
@@ -162,12 +176,34 @@ inline void onMapperAssociationDone(
         return;
     }
 
-    dbus::utility::getAllProperties(
-        connectionName, pcieSlotPath,
-        "xyz.openbmc_project.Inventory.Item.PCIeSlot",
-        [asyncResp](const boost::system::error_code& ec2,
-                    const dbus::utility::DBusPropertiesMap& propertiesList) {
-            onPcieSlotGetAllDone(asyncResp, ec2, propertiesList);
+    sdbusplus::message::object_path invPath("/xyz/openbmc_project/inventory");
+    dbus::utility::getManagedObjects(
+        connectionName, invPath,
+        [asyncResp, connectionName,
+         pcieSlotPath](const boost::system::error_code& ec1,
+                       const dbus::utility::ManagedObjectType& pcieSlotObjs) {
+            if (ec1)
+            {
+                messages::internalError(asyncResp->res);
+                return;
+            }
+            auto& slots = asyncResp->res.jsonValue["Slots"];
+
+            const auto pcieSlotIt = std::ranges::find_if(
+                pcieSlotObjs,
+                [pcieSlotPath](
+                    const std::pair<sdbusplus::message::object_path,
+                                    dbus::utility::DBusInterfacesMap>&
+                        pcieSlot) { return pcieSlotPath == pcieSlot.first; });
+
+            if (pcieSlotIt == pcieSlotObjs.end())
+            {
+                messages::resourceNotFound(asyncResp->res, "PCIeSlots",
+                                           "PCIeSlots_To_Be_Changed");
+                return;
+            }
+
+            processPcieSlot(pcieSlotIt->second, slots);
         });
 }
 
diff --git a/redfish-core/src/redfish.cpp b/redfish-core/src/redfish.cpp
index 7bc67608a7..f550a9ab51 100644
--- a/redfish-core/src/redfish.cpp
+++ b/redfish-core/src/redfish.cpp
@@ -35,6 +35,7 @@
 #include "odata.hpp"
 #include "openbmc/openbmc_managers.hpp"
 #include "pcie.hpp"
+#include "pcie_slots.hpp"
 #include "power.hpp"
 #include "power_subsystem.hpp"
 #include "power_supply.hpp"
@@ -107,6 +108,7 @@ RedfishService::RedfishService(App& app)
     requestRoutesChassis(app);
     requestRoutesChassisResetAction(app);
     requestRoutesChassisResetActionInfo(app);
+    requestRoutesPCIeSlots(app);
     requestRoutesUpdateService(app);
     requestRoutesStorage(app);
     requestRoutesStorageController(app);
-- 
2.43.0

